### PR TITLE
Registry: Implement drafts sortition limit

### LINKED ISSUE #157: Registry: Drafting could result in an OOG error
Since we are using a random seed to fetch a number of jurors from the tree and these need to have enough active balance to be selected for a draft, this function may never end. Thus, we may want to cap the number of iterations to avoid having OOG errors. However, limiting the iterations could trigger some consequences. For instance, we may want to handle this from the `Court` contract, which may lead us to start thinking that further rounds could have fewer jurors drafted than previous ones.

### PR DESCRIPTION
**Pull request by facuspagnuolo**
**MERGE AFTER #175**

Fixes #157 

Assuming that there are lower chances to have jurors repeated for a draft, I'm proposing to also change the way data is returned from the registry draft (related to #74). 

1. Here are some numbers comparing gas costs of `jurors-registry-drafting.js`:

**Before**

```
·----------------------------------------------|---------------------------|-----------|-----------------------------·
|     Solc version: 0.5.8+commit.23d335f2      ·  Optimizer enabled: true  ·  Runs: 1  ·  Block limit: 50000000 gas  │
···············································|···························|···········|······························
|  Methods                                     ·              3 gwei/gas               ·       172.39 usd/eth        │
·························|·····················|·············|·············|···········|···············|··············
|  Contract              ·  Method             ·  Min        ·  Max        ·  Avg      ·  # calls      ·  usd (avg)  │
·························|·····················|·············|·············|···········|···············|··············
|  CourtMockForRegistry  ·  draft              ·      41389  ·     187801  ·   118117  ·          112  ·       0.06  │
·························|·····················|·············|·············|···········|···············|··············
```

**After**

```
·----------------------------------------------|---------------------------|-----------|-----------------------------·
|     Solc version: 0.5.8+commit.23d335f2      ·  Optimizer enabled: true  ·  Runs: 1  ·  Block limit: 50000000 gas  │
···············································|···························|···········|······························
|  Methods                                     ·              3 gwei/gas               ·       172.39 usd/eth        │
·························|·····················|·············|·············|···········|···············|··············
|  Contract              ·  Method             ·  Min        ·  Max        ·  Avg      ·  # calls      ·  usd (avg)  │
·························|·····················|·············|·············|···········|···············|··············
|  CourtMockForRegistry  ·  draft              ·      40351  ·     378957  ·   133770  ·          128  ·       0.07  │
·························|·····················|·············|·············|···········|···············|··············
```

2. Here are some numbers comparing the gas costs of `court-draft.js`: 

**Before**

```
·--------------------------------------------|---------------------------|-----------|-----------------------------·
|    Solc version: 0.5.8+commit.23d335f2     ·  Optimizer enabled: true  ·  Runs: 1  ·  Block limit: 50000000 gas  │
·············································|···························|···········|······························
|  Methods                                   ·              3 gwei/gas               ·       172.61 usd/eth        │
······················|······················|·············|·············|···········|···············|··············
|  Contract           ·  Method              ·  Min        ·  Max        ·  Avg      ·  # calls      ·  usd (avg)  │
······················|······················|·············|·············|···········|···············|··············
|  CourtMock          ·  createDispute       ·     233659  ·     233787  ·   233780  ·          259  ·       0.12  │
······················|······················|·············|·············|···········|···············|··············
|  CourtMock          ·  draft               ·     117717  ·     465032  ·   303078  ·          542  ·       0.16  │
······················|······················|·············|·············|···········|···············|··············
|  CourtMock          ·  heartbeat           ·      89403  ·     667668  ·   255508  ·          642  ·       0.13  │
······················|······················|·············|·············|···········|···············|··············
```

**After**

```
·--------------------------------------------|---------------------------|-----------|-----------------------------·
|    Solc version: 0.5.8+commit.23d335f2     ·  Optimizer enabled: true  ·  Runs: 1  ·  Block limit: 50000000 gas  │
·············································|···························|···········|······························
|  Methods                                   ·              3 gwei/gas               ·       172.61 usd/eth        │
······················|······················|·············|·············|···········|···············|··············
|  Contract           ·  Method              ·  Min        ·  Max        ·  Avg      ·  # calls      ·  usd (avg)  │
······················|······················|·············|·············|···········|···············|··············
|  CourtMock          ·  createDispute       ·     233659  ·     233787  ·   233780  ·          259  ·       0.12  │
······················|······················|·············|·············|···········|···············|··············
|  CourtMock          ·  draft               ·     117106  ·     468961  ·   303837  ·          542  ·       0.16  │
······················|······················|·············|·············|···········|···············|··············
|  CourtMock          ·  heartbeat           ·      89403  ·     667668  ·   255508  ·          642  ·       0.13  │
······················|······················|·············|·············|···········|···············|··············
```

**Conclusion**

In comparison nr. 1, min and avg values are approximated, while the max gas cost value for the proposed changed is actually being doubled. However, these tests are fairly unrealistic, we are forcing several repeated jurors and, in fact, more tests were added in this PR and we can tell that by the difference in the number of calls. 

OTOH, in comparison nr. 2, draft is increased a little bit in ~800 gas. As you can see, these tests haven't been modified in this PR.

### CODE REVIEW

File: test/helpers/registry.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
 const { bn } = require('./numbers')
-const { soliditySha3 } = require("web3-utils");
+const { soliditySha3, toBN } = require('web3-utils')
 
 const expectedBounds = ({ selectedJurors, batchRequestedJurors, balances, totalRequestedJurors }) => {
-  const totalBalance = balances.reduce((acc, x) => acc.add(x), bn(0))
+  const totalBalance = balances.reduce((total, balance) => total.add(balance), bn(0))
```
Comment (bingen) [inline test/helpers/registry.js]: I don't see how this change is related to this PR, and [you already know](https://github.com/aragon/aragon-court/pull/160#discussion_r333385305) I liked it the way it was, but I'm fine with the change.

File: test/helpers/registry.js [hunk 2]
```diff
@@ -20,18 +20,20 @@ const simulateComputeSearchRandomBalances = ({
   let expectedSumTreeBalances = []
   const interval = highActiveBalanceBatchBound.sub(lowActiveBalanceBatchBound)
   for(let i = 0; i < batchRequestedJurors; i++) {
-    const seed = soliditySha3(termRandomness, disputeId, sortitionIteration, i)
-    const balance = bn(lowActiveBalanceBatchBound).add(web3.utils.toBN(seed).mod(interval))
-    expectedSumTreeBalances.push(balance)
+    if (interval.eq(bn(0))) expectedSumTreeBalances.push(lowActiveBalanceBatchBound)
```
Comment (bingen) [inline test/helpers/registry.js]: See my comment above about `interval` zero.

File: contracts/lib/JurorsTreeSortition.sol [hunk 3]
```diff
@@ -119,6 +117,14 @@ library JurorsTreeSortition {
         uint256 interval = _highBatchBound - _lowBatchBound;
         uint256[] memory balances = new uint256[](_batchRequestedJurors);
 
+        // If the given interval is zero, we don't need to compute a random search, simply fill the resulting array with the unique bound
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply contracts/lib/JurorsTreeSortition.sol]: I agree on that from the `Court` contract perspective, however, this is not trivial in this module, we shouldn't make assumptions on one module based on another module's assumptions, even more if these can be changed.

Note that if something changes in the future in the `Court` making this a valid scenario, we will have an invalid opcode at the end. Given that having this check is not harmful (`ISZERO` + `MLOAD` opcode ~= 6 gas) I prefer handling this scenario. WDYT?

File: test/helpers/registry.js (hunk 1 above)
Comment (facuspagnuolo) [reply test/helpers/registry.js]: True, this one wasn't specially related to this PR. However, it was a bit hard to follow some of the logic related to draft simulation we have that I needed to change a bit for this PR. Given that I was updating some of these namings to improve its readability, I made sure to stay consistent in the whole codebase.

File: contracts/lib/JurorsTreeSortition.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (bingen) [reply contracts/lib/JurorsTreeSortition.sol]: Ok, but if you want to keep the check, I think an empty array should be returned.

[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply contracts/lib/JurorsTreeSortition.sol]: Sorry I didn't answer this before

> if `interval` is zero (which means `_batchRequestedJurors` is zero), then it should return an empty array.

Sure, but only if `_batchRequestedJurors` is zero. Note that the active balance in the tree may be zero.

[the remark this answers is not in the document]
Comment (bingen) [reply contracts/lib/JurorsTreeSortition.sol]: Do you mean the total active balance? Well, I don't think we need to put a special check for that: 

- Trying to draft jurors from an empty tree would be quite weird, I wouldn't expect any reasonable outcome from that.
- If the tree is empty, the search will revert anyway:
https://github.com/aragon/aragon-court/blob/development/contracts/lib/HexSumTree.sol#L173
- That lower bound that is being now returned is zero, that doesn't make any sense to me. I also think the return value should be an empty array.
